orttraining/orttraining/python/training/_ortmodule_training_manager.py

```diff
@@ -115,14 +126,14 @@ def backward(ctx, *grad_outputs):
                     elif not grad_output.is_contiguous():
                         grad_output = grad_output.contiguous()
                     contiguous_grad_outputs.append(grad_output)
-                backward_grad_output_ortvalue = [_utils._ortvalue_from_torch_tensor(
-                    grad_output) for grad_output in contiguous_grad_outputs]
 
                 # Run and get results
-                run_id = ctx.run_info.run_id
-                training_io_binding = ctx.run_info.io_binding
-                self._execution_agent.run_backward(backward_grad_output_ortvalue, run_id)
-                backward_outputs = training_io_binding.get_outputs()
+                backward_inputs = C.OrtValueVector()
+                for idx, input in enumerate(contiguous_grad_outputs):
+                    backward_inputs.append(_utils._ortvalue_from_torch_tensor(input))
+
+                backward_outputs = C.OrtValueVector()
+                self._execution_agent.run_backward(backward_inputs, backward_outputs, ctx.run_info.state)
```

Comment:
Could `run_backward()` construct and return the `backward_outputs` itself?

Reply:
that was the case before but I'm a little paranoid about any unnecessary copy taking place in PyBind due to return :) If you insist I can put it back as I do not have any stong objection just felt more comfortable passing opaque types though returning values is probably cleaner. 

---
In reply to: [613575463](https://github.com/microsoft/onnxruntime/pull/7324#discussion_r613575463) [](ancestors = 613575463)

Reply:
Ah, I forgot that `run_backward()` was a pybind-generated method. I'm fine keeping this as it is then. (I suspect that you wouldn't get an extra copy if the code returned the vector thanks to the [RVO](https://en.wikipedia.org/wiki/Copy_elision#Return_value_optimization), but I might be being overly optimistic about the optimizations in pybind, and there might be reasons why this doesn't just work :).)